src/types/internal/schema-field.ts

```diff
@@ -0,0 +1,335 @@
+import {EntityType, ReferenceType, Type} from './type.js';
+import {AnnotationRef} from '../../runtime/arcs-types/annotation.js';
+import {SchemaPrimitiveTypeValue, KotlinPrimitiveTypeValue, BinaryExpressionNode} from '../../runtime/manifest-ast-types/manifest-ast-nodes.js';
+
+export enum Kind {
+  Primitive = 'schema-primitive',
+  KotlinPrimitive = 'kotlin-primitive',
+  Collection = 'schema-collection',
+  Reference = 'schema-reference',
+  OrderedList = 'schema-ordered-list',
+  Union = 'schema-union',
+  Tuple = 'schema-tuple',
+  Nested = 'schema-nested',
+  Inline = 'schema-inline',
+  TypeName = 'type-name' // same as inline.
+}
+
+// tslint:disable-next-line: no-any
+type SchemaFieldMethod  = (field: {}) => FieldType;
+
+export abstract class FieldType {
+  public refinement: Refinement = null;
+  public readonly annotations: AnnotationRef[] = [];
+
+  protected constructor(public readonly kind: Kind) {}
+  get isPrimitive(): boolean { return this.kind === Kind.Primitive; }
+  get isKotlinPrimitive(): boolean { return this.kind === Kind.KotlinPrimitive; }
+  get isCollection(): boolean { return this.kind === Kind.Collection; }
+  get isReference(): boolean { return this.kind === Kind.Reference; }
+  get isOrderedList(): boolean { return this.kind === Kind.OrderedList; }
+  get isUnion(): boolean { return this.kind === Kind.Union; }
+  get isTuple(): boolean { return this.kind === Kind.Tuple; }
+  get isNested(): boolean { return this.kind === Kind.Nested; }
+  get isInline(): boolean { return this.kind === Kind.Inline || this.kind === Kind.TypeName; }
+
+  getType(): SchemaPrimitiveTypeValue | KotlinPrimitiveTypeValue { return null; }
+  getFieldTypes(): FieldType[] { return null; }
+  getFieldType(): FieldType { return null; }
+  getEntityType(): EntityType { return null; }
+
+  abstract toString(): string;
+
+  abstract normalizeForHash(): string;
+
+  clone(): FieldType {
+    // tslint:disable-next-line: no-any
+    const literal: any = this.toLiteral();
+    if (literal.refinement) {
+      literal.refinement = Refinement.fromLiteral(literal.refinement);
+    }
+    return FieldType.fromLiteral(literal);
+  }
+
+  // tslint:disable-next-line: no-any
+  toLiteral(): {} {
+    return {
+      kind: this.kind,
+      annotations: this.annotations,
+      refinement: this.refinement ? this.refinement.toLiteral() : null
+    };
+  }
+
+  equals(other: FieldType): boolean {
+    // TODO(cypher1): structural check instead of stringification.
+    return this.toString() === other.toString();
+  }
+
+  // TODO(shans): output AtLeastAsSpecific here. This is necessary to support
+  // refinements on nested structures and references.
+  isAtLeastAsSpecificAs(other: FieldType): boolean {
+    assert(this.kind === other.kind);
+    return this.equals(other);
+  }
+
+  static create(theField: {}|string): FieldType {
+    let newField = null;
+    // tslint:disable-next-line: no-any
+    const field = theField as any;
```

Comment:
why do you need to do this? You should be able to typeof(theField) == 'string' directly I think.

Reply:
i can do `typeof(theField) == 'string'`, but with {} i cannot then do `field.kind` or `field.refinement` etc.